File: urbanairship-core/src/main/java/com/urbanairship/actions/AddCustomEventAction.java
/* Copyright Airship and Contributors */

package com.urbanairship.actions;

import com.urbanairship.Logger;
import com.urbanairship.analytics.CustomEvent;
import com.urbanairship.json.JsonMap;
import com.urbanairship.push.PushMessage;
import com.urbanairship.util.Checks;

import androidx.annotation.NonNull;

/**
 * An action that adds a custom event.
 * <p>
 * Accepted situations: all
 * <p>
 * Accepted argument value - A map of fields for the custom event:
 * <ul>
 * <li>{@link com.urbanairship.analytics.CustomEvent#EVENT_NAME}: String, Required</li>
 * <li>{@link com.urbanairship.analytics.CustomEvent#EVENT_VALUE}: number as a String or Number</li>
 * <li>{@link com.urbanairship.analytics.CustomEvent#TRANSACTION_ID}: String</li>
 * <li>{@link com.urbanairship.analytics.CustomEvent#INTERACTION_ID}: String</li>
 * <li>{@link com.urbanairship.analytics.CustomEvent#INTERACTION_TYPE}: String</li>
 * <li>{@link com.urbanairship.analytics.CustomEvent#PROPERTIES}: JsonMap of Strings, Booleans, Numbers, or arrays of Strings</li>
 * </ul>
 * When a custom event action is triggered from a Message Center Rich Push Message, the interaction type
 * and ID will automatically be filled for the message if they are left blank.
 * <p>
 * Result value: <code>null</code>
 * <p>
 * Default Registration Name: add_custom_event_action
 * <p>
 * Default Registration Predicate: Rejects SITUATION_PUSH_RECEIVED
 */
public class AddCustomEventAction extends Action {

    /**
     * Default registry name
     */
    @NonNull
    public static final String DEFAULT_REGISTRY_NAME = "add_custom_event_action";

    @NonNull
    @Override
    public ActionResult perform(@NonNull ActionArguments arguments) {
        JsonMap customEventMap = arguments.getValue().toJsonValue().optMap();

        // Parse the event values from the map
        String eventName = customEventMap.opt(CustomEvent.EVENT_NAME).getString();
        Checks.checkNotNull(eventName, "Missing event name");

        String eventStringValue = customEventMap.opt(CustomEvent.EVENT_VALUE).getString();
        double eventDoubleValue = customEventMap.opt(CustomEvent.EVENT_VALUE).getDouble(0);

        String transactionId = customEventMap.opt(CustomEvent.TRANSACTION_ID).getString();
        String interactionType = customEventMap.opt(CustomEvent.INTERACTION_TYPE).getString();
        String interactionId = customEventMap.opt(CustomEvent.INTERACTION_ID).getString();
        JsonMap properties = customEventMap.opt(CustomEvent.PROPERTIES).getMap();

        CustomEvent.Builder eventBuilder = CustomEvent.newBuilder(eventName)
                .setTransactionId(transactionId)
                .setAttribution((PushMessage) arguments.getMetadata().getParcelable(ActionArguments.PUSH_MESSAGE_METADATA))
                .setInteraction(interactionType, interactionId);

        if (eventStringValue != null) {
            eventBuilder.setEventValue(eventStringValue);
        } else {
            eventBuilder.setEventValue(eventDoubleValue);
        }

        // Try to fill in the interaction if its not set
        if (interactionId == null && interactionType == null) {
            String messageId = arguments.getMetadata().getString(ActionArguments.RICH_PUSH_ID_METADATA);

            if (messageId != null) {
                eventBuilder.setMessageCenterInteraction(messageId);
            }
        }

        if (properties != null) {
            eventBuilder.setProperties(properties);
        }

        CustomEvent event = eventBuilder.build();
        event.track();
        if (event.isValid()) {
            return ActionResult.newEmptyResult();
        } else {
            return ActionResult.newErrorResult(new IllegalArgumentException("Unable to add custom event. Event is invalid."));
        }
    }

    @Override
    public boolean acceptsArguments(@NonNull ActionArguments arguments) {
        if (arguments.getValue().getMap() != null) {
            if (arguments.getValue().getMap().get("event_name") == null) {
                Logger.error("CustomEventAction requires an event name in the event data.");
                return false;
            }
            return true;
        }

        Logger.error("CustomEventAction requires a map of event data.");
        return false;
    }

    /**
     * Default {@link AddCustomEventAction} predicate.
     */
    public static class AddCustomEventActionPredicate implements ActionRegistry.Predicate {

        @Override
        public boolean apply(@NonNull ActionArguments arguments) {
            return Action.SITUATION_PUSH_RECEIVED != arguments.getSituation();
        }

    }

}

File: urbanairship-core/src/main/java/com/urbanairship/actions/AddCustomEventAction.kt
/* Copyright Airship and Contributors */
package com.urbanairship.actions

import android.os.Parcelable
import androidx.annotation.RestrictTo
import com.urbanairship.Airship
import com.urbanairship.UALog
import com.urbanairship.actions.ActionResult.Companion.newEmptyResult
import com.urbanairship.actions.ActionResult.Companion.newErrorResult
import com.urbanairship.analytics.CustomEvent
import com.urbanairship.analytics.CustomEvent.Companion.newBuilder
import com.urbanairship.json.JsonMap
import com.urbanairship.json.JsonValue
import com.urbanairship.json.optionalField
import com.urbanairship.json.optionalMap
import com.urbanairship.push.PushMessage

/**
 * An action that adds a custom event.
 *
 *
 * Accepted situations: all
 *
 *
 * Accepted argument value - A map of fields for the custom event:
 *
 *  * [com.urbanairship.analytics.CustomEvent.EVENT_NAME]: String, Required
 *  * [com.urbanairship.analytics.CustomEvent.EVENT_VALUE]: number as a String or Number
 *  * [com.urbanairship.analytics.CustomEvent.TRANSACTION_ID]: String
 *  * [com.urbanairship.analytics.CustomEvent.INTERACTION_ID]: String
 *  * [com.urbanairship.analytics.CustomEvent.INTERACTION_TYPE]: String
 *  * [com.urbanairship.analytics.CustomEvent.PROPERTIES]: JsonMap of Strings, Booleans, Numbers, or arrays of Strings
 *
 * When a custom event action is triggered from a Message Center Rich Push Message, the interaction type
 * and ID will automatically be filled for the message if they are left blank.
 *
 *
 * Result value: `null`
 *
 *
 * Default Registration Name: [DEFAULT_NAMES]
 *
 *
 * Default Registration Predicate: Rejects [Action.Situation.PUSH_RECEIVED]
 */
public class AddCustomEventAction(
    private val eventRecord: (CustomEvent) -> Unit = {
        Airship.analytics.recordCustomEvent(it)
    }
) : Action() {

    override fun perform(arguments: ActionArguments): ActionResult {
        val customEventMap = arguments.value.toJsonValue().optMap()

        // Parse the event values from the map
        val eventName = getEventName(customEventMap)
        requireNotNull(eventName) { "Missing event name" }

        val eventDoubleValue: Double = getEventValue(customEventMap) ?: 0.0

        val interactionType = customEventMap.optionalField<String>(CustomEvent.INTERACTION_TYPE)
        val interactionId = customEventMap.optionalField<String>(CustomEvent.INTERACTION_ID)

        val eventBuilder = newBuilder(eventName)
            .setTransactionId(customEventMap.optionalField(CustomEvent.TRANSACTION_ID))
            .setAttribution(arguments.metadata.getParcelable<Parcelable>(ActionArguments.PUSH_MESSAGE_METADATA) as PushMessage?)
            .setInteraction(interactionType, interactionId)

        eventBuilder.setEventValue(eventDoubleValue)

        arguments.metadata.getString(IN_APP_CONTEXT_METADATA_KEY)?.let { inApp ->
            try {
                eventBuilder.setInAppContext(JsonValue.parseString(inApp))
            } catch (e: Exception) {
                UALog.w("Failed to parse in-app context for custom event", e)
            }
        }

        // Try to fill in the interaction if its not set
        if (interactionId == null && interactionType == null) {
            arguments.metadata.getString(ActionArguments.RICH_PUSH_ID_METADATA)?.let {
                eventBuilder.setMessageCenterInteraction(it)
            }
        }

        eventBuilder.setProperties(customEventMap.optionalMap(CustomEvent.PROPERTIES))

        val event = eventBuilder.build()

        return if (event.isValid()) {
            eventRecord(event)
            newEmptyResult()
        } else {
            newErrorResult(IllegalArgumentException("Unable to add custom event. Event is invalid."))
        }
    }

    override fun acceptsArguments(arguments: ActionArguments): Boolean {
        val map = arguments.value.map ?: run {
            UALog.e("CustomEventAction requires a map of event data.")
            return false
        }

        if (!map.containsKey(KEY_NAME) && !map.containsKey(CustomEvent.EVENT_NAME)) {
            UALog.e("CustomEventAction requires an event name in the event data.")
            return false
        }

        return true
    }

    private fun getEventName(map: JsonMap): String? {
        return map.optionalField(KEY_NAME)
            ?: map.optionalField(CustomEvent.EVENT_NAME)
    }

    private fun getEventValue(map: JsonMap): Double? {
        return map.optionalField<String>(KEY_VALUE)?.toDoubleOrNull()
            ?: map.optionalField<Double>(KEY_VALUE)
            ?: map.optionalField<Double>(CustomEvent.EVENT_VALUE)
    }

    /**
     * Default [AddCustomEventAction] predicate.
     */
    public class AddCustomEventActionPredicate public constructor() : ActionPredicate {

        override fun apply(arguments: ActionArguments): Boolean {
            return Situation.PUSH_RECEIVED != arguments.situation
        }
    }

    public companion object {

        internal const val KEY_NAME: String = "name"
        internal const val KEY_VALUE: String = "value"

        /**
         * Default action names.
         */
        public val DEFAULT_NAMES: Set<String> = setOf("add_custom_event_action", "^+e")

        /**
         * @hide
         */
        @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
        public const val IN_APP_CONTEXT_METADATA_KEY: String = "in_app_metadata"
    }
}
